README.md
# Dotenv Editor
This tool makes it easier for users to setup and configure an application
which uses a dotenv file for its configuration.

![dotenv-editor-screenshot](https://user-images.githubusercontent.com/503929/91663533-bd27c080-eae9-11ea-9d1d-9e2196b23b11.png)

Dotenv files (https://en.wikipedia.org/wiki/Env) are used by docker (compose).
A lot of programming languages have libraries to handle .env files:
* [Node.js](https://github.com/motdotla/dotenv)
* [Go](https://github.com/joho/godotenv)
* [Python](https://github.com/theskumar/python-dotenv)
* [Ruby](https://github.com/bkeepers/dotenv)
* [Java](https://github.com/cdimascio/java-dotenv)
* [PHP](https://github.com/vlucas/phpdotenv)


### Usage
Start the application by running the script:

```
./dotenv-editor
```

It will display a list of all variables defined in the existing .env file (or
.env.sample file). It tries to group elements by prefix to improve user
experience in case of larger or more complex .env files.


### Options
```
Usage: dotenv-editor [FILE]

Command options:
  -f, --file=FILE            use this dotenv configuration file (default: .env)
  -l, --log=FILE             use this file for (debug) logging (default: /dev/null)
  -v, --version              print program version
  -h, --help                 print help info (this screen)

Compatibility modes:
  -q, --quote                quote variables with a '"'

UI features:
  -g, --group                enable grouping variables by separator
  -c, --checkbox             enable checkboxes for variables called *ENABLED* or
                             have a 0 or 1 value

Specify UI library:
  --dialog                   Use the original dialog library
  --whiptail                 Use the newer whiptail library
```

If no file parameter (-f) is provided it wil look like for the .env file. In
case the .env does not exists it wil check for any .env-* file to use as an
example to start from after which it will be copied to the .env file.


### Examples
The examples directory contains a couple of example .env files to show the
features of the dotenv-editor script.

#### Simple example
The most basic example contains of a (small) list of key/value pairs. This
works good if the amount of variables to configure is not too large.
```
./dotenv-editor -f=example/simple
```

#### Grouped example
To configure a large list of variables (key/value pairs) the menu items can be
grouped using the *-g option*. This option will split the variables on an
underscore. (Leave out the -g option to see the difference)
```
./dotenv-editor -f=example/grouped -g
```

#### Checkboxes example
Enable/disable a group of variables can be used with the *-c option*. A service
is a variabled ending with "$NAME_ENABLED". In case the value of this variable will
be set to 0 all variables starting with $NAME will be hidden.
```
./dotenv-editor -f=example/checked -c
```

### Install
The script can easily be installed globally by running:
```
make install
```
Run *make* to see a list of all possible targets.


### Requirements
The dotenv-editor is a stand-alone tool and only uses basic unix tools (such as
bash, grep, awk). The only requirement it to have at least one CLI UI rendering
tool installed (such as dialog or whiptail). The script will detect which
library is available or ask you to install one if none found.

### Configuration
The default options can be changed by creating a ~/.dotenv-editor file. This
file contains variable defaults.
```
MODE_GROUP="ON"
MODE_CHECKBOX="OFF"
COMPAT_QUOTE="ON"
LOG_FILE=/tmp/dotenv-editor.log
```

### Contributing
In case you experience a bug or have an improvement please fork the git repository and push a git merge requests.

In case you have an idea but you are unable to implement it, then add your idea in the bottom section (future versions) of the [TODO.md](TODO.md) file.

Once a feature/bug has been implemented/fixed it can be merged/pushed into the master branch. Once a release is made the implemented items from the [TODO](TODO.md) list will be moved to the [CHANGELOG](CHANGELOG.md) file.


### Vision
The goal for this project is to be a
* simple tool (user-friendly)
* using common unix tools (no other dependencies)
* follow the .env file specification (use standard specs)


### Credits
The following people get a free beer next time I see them in a bar as they contributed to this project:
* [Eric Herman](https://github.com/ericherman)


### License
Licensed under the terms of the MIT License.

See [LICENSE](LICENSE) for details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.


## 0.5.0 (2020-10-04)
Enable features by default and can be disabled using one of the 
--disable-<feature> flags./

### Changed
+ Rename option --checkbox => --services
+ Options enabled by default; disable using --disable-<feature> flag.


## 0.4.4 (2020-10-04)
More command line options for tweaking behaviour.

### Added
- Option (--depth=2) to define the group depth level.
- Option (--modify) for enabling/disabling added/deleting variables.


## 0.4.3 (2020-09-28)
Show a progress bar loading a larger files.

### Added
+ Show progress bar when progressing files (improve performance later)

### Fixed
+ Ignore files starting with a space (fixes loading laradock example)


## 0.4.2 (2020-09-27)
Various bugfixes and improvements.

### Fixed
+ Description/value not set when adding a new variable.
+ List is empty after VIEW_DIFF (depth level too high?)
+ View changes after enabling/disabling services


## 0.4.1 (2020-09-27)
Descriptions for grouped items

### Added
+ Extract a description for the group. (default: 'More options')

### Improvements
+ Improved README (& added github pages)


## 0.4.0 (2020-09-19)
Supports description of variables using comments.

### Improvements
+ Extract variable comment by getting the previous lines in the .env file which
  start with a #. The comment is used to give more information how a variable
  is used and changes the applications behaviour.

### Fixes
+ Go back to previous depth level after editing a variable
+ Hide ENABLED fields when the services checkbox feature is enabled


## 0.3.5 (2020-09-13)
Various small fixes and code clean up

### Fixes
+ Do not quote integer values

### Removed
+ Welcome message


## 0.3.4 (2020-09-10)
Some small fixes and improvements to the README file.

### Added
+ Ask user to create a new file if no .env file can be found

### Fixed
+ Fix running the editor when there is not .env file in the current dir


## 0.3.3 (2020-09-06)
Menu for enable/disable services. Also various fixes and improvements.

### Added
+ Add -e option to group enable variables (*_ENABLED). It display a checkbox
  list with items to select, which will be used to filter upon.
+ Parameter --depth to limit the depth for grouping (enables group mode)
+ Create debian package

### Fixes
+ Cancel button for preferences resets the variables
+ Only make variable name uppercase when creating a new variable
+ Go back to previous menu after using settings, save, reset or new variable
+ Improve log messages (make it more human readable)
+ Displaying menu with items with depth > 2


## 0.3.2 (2020-09-05)
Create and delete variables.

### Added
+ Create new variable
+ Delete variable in case a variable is empty after asking input
+ Makefile for (un)installing
+ Store preferences in ~/.dotenv-editor
+ Use current working directory to search for .env files


## 0.3.1 (2020-09-01)
Main menu for common tasks (reset, save, view changes) and ask before saving changes.

### Added
+ Add main menu with options to change / save preferences, create new item, view diff.
+ Copy .env file to tmp file and ask before saving.
+ Ask to save changes before exiting
+ Show differences in a textbox

### Fixed
+ Bug selecting variable in simple (non-group) mode


## 0.3.0 (2020-08-31)
Group variables by prefix in menus.

### Added
+ Flag to group variables (like the previous version did already). This could
  also be implemented as a maximum depth level to group upon.

### Fixed
+ Do not create a file called 1 (a bug)
+ After asking for input return to the previous menu, not the main menu.


## 0.2.3 (2020-08-30)
Various UI Fixed

### Added
+ Compatibility mode to quote string variables

### Removed
+ Removed screenshot image from the git repo (moved to github pages).

### Fixed
+ Error when variable value contains spaces
+ Improve layout for menu list (more space)


## 0.2.2 (2020-08-25)
Support multiple libraries and various Fixed.

### Added
+ Support multiple libraries (dialog/whiptail/zenity/whatever)
+ Specify log output (to file or use /dev/null by default)
+ Option to specify .env file to use

### Removed
+ Unrelevant log messages
+ Clean up comments


## 0.2.1 (2020-08-24)
Various bugFixed and improvements.

### Added
+ Check if required dependencies (sed, awk, dialog, whiptail, etc) exists.
+ Help option to display the command arguments and options
+ Version flag

### Improvements
+ Show current value in menu for variables

### Removed
+ Drop dependency: bin/dotenv.sh
+ Old proof of concepts scripts (poc/)


## 0.2 (2020-08-23)
Working proof of concept which will update a configuration setting in a dotenv
file.

### Added
- Welcome screen
- Use underscores to split variables to create sub menus
- Updating .env file after asking user input
- Dotenv file selection if there is .env in the current directory


## 0.1 (2020-08-18)
Some first executable prototype. It will display the found variables in a menu
using the dialog/whiptail command.

### Added
- Proof of concept menu using dialog/whiptail


menu_group.sh
#!/bin/bash

source .env

DIALOG=${DIALOG=dialog}

# Loop over all variables with the existing defaults.
VARIABLE_LIST=`awk -F= '{print $1 }' .env  | grep -v '^#' | grep -v '^$'`
VARIABLE_GROUP=`awk -F= '{print $1 }' .env  | grep -v '^#' | grep -v '^$' | awk -F"_" '{print $1}' | sort | uniq`
# VARIABLE_GROUPS=`echo $VARIABLE_LIST | awk -F"_" '{print $1}'`

MENU_LIST=""
for i in $VARIABLE_LIST; do
	PREFIX=`echo $i | awk -F"_" '{print $1}'`
	REST=`echo $i | awk -F"_" '{print $2}'`

    VALUE=`./bin/dotenv.sh get $i`
    MENU_LIST="$MENU_LIST $i \"$VALUE\" "
    # echo "- $i => $VALUE"
    #$DIALOG --clear --title "Configuration - $i" --inputbox "$i" 8 80 "`./bin/dotenv.sh get $i`"
done


echo $MENU_LIST

# $DIALOG
#     --menu "You can use the UP/DOWN arrow keys, the first \n \
#     --clear \
#     --title "[ M A I N - M E N U ]" \
#     letter of the choice as a hot key, or the \n\
#     number keys 1-9 to choose an option.\n\
#     Choose the TASK" 20 60 14 \
    # $MENU_LIST 
# BIER "echo BIER" \
# Main "Main MENU" \
# Sub "Sub MENU" \
# Date/time "Displays date and time" \
# Calendar "Displays a calendar" \
# Editor "Start a text editor" \
	# Exit "Exit to the shell" 2>"${INPUT}"


# Success screen
#$DIALOG --title "Finished" --clear \
#	--yesno "$APP_TITLE\n\nThe application settings have been configured to your needs. The next step is to run:\n\n  make start\n\nTo start the docker container(s)." 20 80



poc/copied_script.sh
#!/bin/bash
# utilitymenu.sh - A sample shell script to display menus on screen
# Store menu options selected by the user
INPUT=/tmp/menu.sh.$$

# Storage file for displaying cal and date command output
OUTPUT=/tmp/output.sh.$$

# get text editor or fall back to vi_editor
vi_editor=${EDITOR-vi}

# trap and delete temp files
trap "rm $OUTPUT; rm $INPUT; exit" SIGHUP SIGINT SIGTERM

#
# Purpose - display output using msgbox 
#  $1 -> set msgbox height
#  $2 -> set msgbox width
#  $3 -> set msgbox title
#
function display_output(){
	local h=${1-10}			# box height default 10
	local w=${2-41} 		# box width default 41
	local t=${3-Output} 	# box title 
	dialog --backtitle "Linux Shell Script Tutorial" --title "${t}" --clear --msgbox "$(<$OUTPUT)" ${h} ${w}
}
#
# Purpose - display current system date & time
#
function show_date(){
	echo "Today is $(date) @ $(hostname -f)." >$OUTPUT
    display_output 6 60 "Date and Time"
}
#
# Purpose - display a calendar
#
function show_calendar(){
	cal >$OUTPUT
	display_output 13 25 "Calendar"
}
#
# set infinite loop
#
while true
do

### display main menu ###
dialog --clear  --help-button --backtitle "Linux Shell Script Tutorial" \
--title "[ M A I N - M E N U ]" \
--menu "You can use the UP/DOWN arrow keys, the first \n\
letter of the choice as a hot key, or the \n\
number keys 1-9 to choose an option.\n\
Choose the TASK" 20 60 14 \
BIER "echo BIER" \
Date/time "Displays date and time" \
Calendar "Displays a calendar" \
Editor "Start a text editor" \
Exit "Exit to the shell" 2>"${INPUT}"

menuitem=$(<"${INPUT}")


# make decsion 
case $menuitem in
	Date/time) show_date;;
	Calendar) show_calendar;;
	Editor) $vi_editor;;
	BIER) echo "BIER!!!!"; sleep 1;;
	Exit) echo "Bye"; break;;
esac

done

# if temp files found, delete em
[ -f $OUTPUT ] && rm $OUTPUT
[ -f $INPUT ] && rm $INPUT


poc/basic_menu.sh
#!/bin/bash

source .env

#DIALOG=${DIALOG=whiptail}
DIALOG=${DIALOG=dialog}

# Intro screen
$DIALOG --clear \
	--title "Setup" --clear \
        --yesno "$APP_TITLE\n\nThis setup will guide you through configuring all the existing variables.\n\nTheir defaults are already prefilled, so hit ENTER to go the next one." 20 80


# utilitymenu.sh - A sample shell script to display menus on screen
# Store menu options selected by the user
INPUT=/tmp/menu.sh.$$

# Storage file for displaying cal and date command output
OUTPUT=/tmp/output.sh.$$

# get text editor or fall back to vi_editor
vi_editor=${EDITOR-vi}

# trap and delete temp files
trap "rm $OUTPUT; rm $INPUT; exit" SIGHUP SIGINT SIGTERM

#
# Purpose - display output using msgbox 
#  $1 -> set msgbox height
#  $2 -> set msgbox width
#  $3 -> set msgbox title
#
function display_output(){
	local h=${1-10}			# box height default 10
	local w=${2-41} 		# box width default 41
	local t=${3-Output} 	# box title 
	$DIALOG --backtitle "Linux Shell Script Tutorial" --title "${t}" --clear --msgbox "$(<$OUTPUT)" ${h} ${w}
}


#
# Purpose - display current system date & time
#
function show_date(){
	echo "Today is $(date) @ $(hostname -f)." >$OUTPUT
    display_output 6 60 "Date and Time"
}
#
# Purpose - display a calendar
#
function show_calendar(){
	cal >$OUTPUT
	display_output 13 25 "Calendar"
}

function menu_main() {
    # Loop over all variables with the existing defaults.
    VARIABLE_LIST=`awk -F= '{print $1 }' .env  | grep -v '^#' | grep -v '^$'`
    for i in $VARIABLE_LIST; do
        VALUE=`./bin/dotenv.sh get $i`
        $DIALOG --clear --title "Configuration - $i" --inputbox "$i" 8 80 "`./bin/dotenv.sh get $i`"
    done

    # Success screen
    $DIALOG --title "Finished" --clear \
            --yesno "$APP_TITLE\n\nThe application settings have been configured to your needs. The next step is to run:\n\n  make start\n\nTo start the docker container(s)." 20 80
}


function menu_sub() {
    # Loop over all variables with the existing defaults.
    VARIABLE_LIST=`awk -F= '{print $1 }' .env  | grep -v '^#' | grep -v '^$'`
    for i in $VARIABLE_LIST; do
        VALUE=`./bin/dotenv.sh get $i`
        $DIALOG --clear --title "Configuration - $i" --inputbox "$i" 8 80 "`./bin/dotenv.sh get $i`"
    done

    # Success screen
    $DIALOG --title "Finished" --clear \
            --yesno "$APP_TITLE\n\nThe application settings have been configured to your needs. The next step is to run:\n\n  make start\n\nTo start the docker container(s)." 20 80
}


#
# set infinite loop
#
while true
do

### display main menu ###
$DIALOG --clear  --help-button --backtitle "Linux Shell Script Tutorial" \
    --title "[ M A I N - M E N U ]" \
    --menu "You can use the UP/DOWN arrow keys, the first \n\
    letter of the choice as a hot key, or the \n\
    number keys 1-9 to choose an option.\n\
    Choose the TASK" 20 60 14 \
BIER "echo BIER" \
Main "Main MENU" \
Sub "Sub MENU" \
Date/time "Displays date and time" \
Calendar "Displays a calendar" \
Editor "Start a text editor" \
Exit "Exit to the shell" 2>"${INPUT}"

menuitem=$(<"${INPUT}")


# make decsion 
case $menuitem in
	Date/time) show_date;;
	Calendar) show_calendar;;
	Main) menu_main;;
	Sub) menu_sub;;
	Editor) $vi_editor;;
	BIER) echo "BIER!!!!"; sleep 1;;
	Exit) echo "Bye"; clear; break;;
esac

done

# if temp files found, delete em
[ -f $OUTPUT ] && rm $OUTPUT
[ -f $INPUT ] && rm $INPUT

